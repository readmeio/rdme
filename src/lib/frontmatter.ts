import type { Mappings } from './readmeAPIFetch.js';
import type { PageMetadata } from './readPage.js';
import type { CommandClass } from '../index.js';
import type { ErrorObject } from 'ajv';
import type { SchemaObject } from 'oas/types';

import fs from 'node:fs';

import { Ajv } from 'ajv';
import _addFormats from 'ajv-formats';
import grayMatter from 'gray-matter';

// workaround from here: https://github.com/ajv-validator/ajv-formats/issues/85#issuecomment-2262652443
const addFormats = _addFormats as unknown as typeof _addFormats.default;

/**
 * Validates the frontmatter data, fixes any issues, and returns the results.
 */
export function fix(
  this: CommandClass['prototype'],
  /** frontmatter data to be validated */
  data: PageMetadata['data'],
  /** schema to validate against */
  schema: SchemaObject,
  /**
   * mappings of object IDs to slugs
   * (e.g., category IDs to category URIs)
   */
  mappings: Mappings,
): {
  changeCount: number;
  errors: ErrorObject[];
  hasIssues: boolean;
  unfixableErrors: ErrorObject[];
  updatedData: PageMetadata['data'];
} {
  if (!Object.keys(data).length) {
    this.debug('no frontmatter attributes found, skipping validation');
    return { changeCount: 0, errors: [], hasIssues: false, unfixableErrors: [], updatedData: data };
  }

  const ajv = new Ajv({ allErrors: true, strictTypes: false, strictTuples: false });

  addFormats(ajv);

  const ajvValidate = ajv.compile(schema);

  ajvValidate(data);

  let errors: ErrorObject[] | undefined;

  if (ajvValidate?.errors) {
    errors = ajvValidate.errors;
    this.debug(`${errors.length} errors found: ${JSON.stringify(errors)}`);

    // if one of the errors is that the category uri doesn't match the pattern,
    // we can ignore it since we normalize it later
    const categoryUriPatternErrorIndex = ajvValidate.errors.findIndex(
      error => error.instancePath === '/category/uri' && error.keyword === 'pattern',
    );

    if (categoryUriPatternErrorIndex >= 0) {
      this.debug('removing category uri pattern error');
      errors.splice(categoryUriPatternErrorIndex, 1);
    }

    // also do the same for the parent uri
    const parentUriPatternErrorIndex = ajvValidate.errors.findIndex(
      error => error.instancePath === '/parent/uri' && error.keyword === 'pattern',
    );

    if (parentUriPatternErrorIndex >= 0) {
      this.debug('removing category uri pattern error');
      errors.splice(parentUriPatternErrorIndex, 1);
    }
  }

  let changeCount = 0;
  const unfixableErrors: ErrorObject[] = [];
  const updatedData = structuredClone(data);

  if (typeof errors === 'undefined' || !errors.length) {
    return { changeCount, errors: [], hasIssues: false, unfixableErrors, updatedData };
  }

  errors.forEach(error => {
    if (error.instancePath === '/category' && error.keyword === 'type') {
      const uri = mappings.categories[data.category as string];
      updatedData.category = {
        uri: uri || `uri-that-does-not-map-to-${data.category}`,
      };
      changeCount += 1;
    } else if (error.keyword === 'additionalProperties') {
      const badKey = error.params.additionalProperty as string;
      const extractedValue = data[badKey];
      if (error.schemaPath === '#/additionalProperties') {
        // if the bad property is at the root level, delete it
        delete updatedData[badKey];
        changeCount += 1;
        if (badKey === 'excerpt') {
          // if the `content` object exists, add to it. otherwise, create it
          if (typeof updatedData.content === 'object' && updatedData.content) {
            (updatedData.content as Record<string, unknown>).excerpt = extractedValue;
          } else {
            updatedData.content = {
              excerpt: extractedValue,
            };
          }
        } else if (badKey === 'categorySlug') {
          updatedData.category = {
            uri: extractedValue,
          };
        } else if (badKey === 'parentDoc') {
          const uri = mappings.parentPages[extractedValue as string];
          if (uri) {
            updatedData.parent = {
              uri,
            };
          }
        } else if (badKey === 'parentDocSlug') {
          updatedData.parent = {
            uri: extractedValue,
          };
        } else if (badKey === 'hidden') {
          updatedData.privacy = { view: extractedValue ? 'anyone_with_link' : 'public' };
        } else if (badKey === 'order') {
          updatedData.position = extractedValue;
        }
      } else {
        unfixableErrors.push(error);
      }
    } else {
      unfixableErrors.push(error);
    }
  });

  return { errors, changeCount, hasIssues: true, unfixableErrors, updatedData };
}

export function writeFixes(
  this: CommandClass['prototype'],
  /** all metadata for the page that will be written to */
  metadata: PageMetadata,
  /** frontmatter changes to be applied */
  updatedData: PageMetadata['data'],
) {
  this.debug(`writing fixes to ${metadata.filePath}`);
  const result = grayMatter.stringify(metadata.content, updatedData);
  fs.writeFileSync(metadata.filePath, result, { encoding: 'utf-8' });

  const updatedMetadata = {
    ...metadata,
    data: updatedData,
  };

  return updatedMetadata;
}
