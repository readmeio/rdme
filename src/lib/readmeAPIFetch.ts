import type { SpecFileType } from './prepareOas';
import type { RequestInit, Response } from 'node-fetch';

import path from 'path';

import config from 'config';
import mime from 'mime-types';
import nodeFetch, { Headers } from 'node-fetch'; // eslint-disable-line no-restricted-imports

import pkg from '../../package.json';

import APIError from './apiError';
import { git } from './createGHA';
import isCI, { ciName, isGHA } from './isCI';
import { debug, warn } from './logger';

const SUCCESS_NO_CONTENT = 204;

/**
 * This contains a few pieces of information about a file so
 * we can properly construct a source URL for it.
 */
interface FilePathDetails {
  /** The URL or local file path */
  filePath: string;
  /** This is derived from the `oas-normalize` `type` property. */
  fileType: SpecFileType;
}

function getProxy() {
  // this is something of an industry standard env var, hence the checks for different casings
  const proxy = process.env.HTTPS_PROXY || process.env.https_proxy;
  if (proxy) {
    // adds trailing slash
    return proxy.endsWith('/') ? proxy : `${proxy}/`;
  }
  return '';
}

/**
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Warning}
 * @see {@link https://www.rfc-editor.org/rfc/rfc7234#section-5.5}
 * @see {@link https://github.com/marcbachmann/warning-header-parser}
 */
interface WarningHeader {
  agent: string;
  code: string;
  date?: string;
  message: string;
}

function stripQuotes(s: string) {
  if (!s) return '';
  return s.replace(/(^"|[",]*$)/g, '');
}

/**
 * Parses Warning header into an array of warning header objects
 * @param header raw `Warning` header
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Warning}
 * @see {@link https://www.rfc-editor.org/rfc/rfc7234#section-5.5}
 * @see {@link https://github.com/marcbachmann/warning-header-parser}
 */
function parseWarningHeader(header: string): WarningHeader[] {
  try {
    const warnings = header.split(/([0-9]{3} [a-z0-9.@\-/]*) /g);

    let previous: WarningHeader;

    return warnings.reduce((all, w) => {
      // eslint-disable-next-line no-param-reassign
      w = w.trim();
      const newError = w.match(/^([0-9]{3}) (.*)/);
      if (newError) {
        previous = { code: newError[1], agent: newError[2], message: '' };
      } else if (w) {
        const errorContent = w.split(/" "/);
        if (errorContent) {
          previous.message = stripQuotes(errorContent[0]);
          previous.date = stripQuotes(errorContent[1]);
          all.push(previous);
        }
      }
      return all;
    }, []);
  } catch (e) {
    debug(`error parsing warning header: ${e.message}`);
    return [{ code: '199', agent: '-', message: header }];
  }
}

/**
 * Getter function for a string to be used in the user-agent header based on the current
 * environment.
 *
 */
function getUserAgent() {
  const gh = isGHA() ? '-github' : '';
  return `rdme${gh}/${pkg.version}`;
}

/**
 * Creates a relative path for the file from the root of the repo,
 * otherwise returns the path
 */
async function normalizeFilePath(opts: FilePathDetails) {
  if (opts.fileType === 'path') {
    const repoRoot = await git.revparse(['--show-toplevel']).catch(e => {
      debug(`[fetch] error grabbing git root: ${e.message}`);
      return '';
    });

    return path.relative(repoRoot, opts.filePath);
  }
  return opts.filePath;
}

/**
 * Sanitizes and stringifies the `Headers` object for logging purposes
 */
function sanitizeHeaders(headers: Headers) {
  const raw = new Headers(headers).raw();
  if (raw.Authorization) raw.Authorization = ['redacted'];
  return JSON.stringify(raw);
}

/**
 * Wrapper for the `fetch` API so we can add rdme-specific headers to all API requests.
 *
 * @param pathname the pathname to make the request to. Must have a leading slash.
 * @param fileOpts optional object containing information about the file being sent.
 * We use this to construct a full source URL for the file.
 */
export default async function readmeAPIFetch(
  pathname: string,
  options: RequestInit = { headers: new Headers() },
  fileOpts: FilePathDetails = { filePath: '', fileType: false },
) {
  let source = 'cli';
  let headers = options.headers as Headers;

  if (!(options.headers instanceof Headers)) {
    headers = new Headers(options.headers);
  }

  headers.set('User-Agent', getUserAgent());

  if (isGHA()) {
    source = 'cli-gh';
    headers.set('x-github-repository', process.env.GITHUB_REPOSITORY);
    headers.set('x-github-run-attempt', process.env.GITHUB_RUN_ATTEMPT);
    headers.set('x-github-run-id', process.env.GITHUB_RUN_ID);
    headers.set('x-github-run-number', process.env.GITHUB_RUN_NUMBER);
    headers.set('x-github-sha', process.env.GITHUB_SHA);

    const filePath = await normalizeFilePath(fileOpts);

    if (filePath) {
      /**
       * Constructs a full URL to the file using GitHub Actions runner variables
       * @see {@link https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables}
       * @example https://github.com/readmeio/rdme/blob/cb4129d5c7b51ff3b50f933a9c7d0c3d0d33d62c/documentation/rdme.md
       */
      try {
        const sourceUrl = new URL(
          `${process.env.GITHUB_SERVER_URL}/${process.env.GITHUB_REPOSITORY}/blob/${process.env.GITHUB_SHA}/${filePath}`,
        ).href;
        headers.set('x-readme-source-url', sourceUrl);
      } catch (e) {
        debug(`error constructing github source url: ${e.message}`);
      }
    }
  }

  if (isCI()) {
    headers.set('x-rdme-ci', ciName());
  }

  headers.set('x-readme-source', source);

  if (fileOpts.filePath && fileOpts.fileType === 'url') {
    headers.set('x-readme-source-url', fileOpts.filePath);
  }

  const fullUrl = `${getProxy()}${config.get('host')}${pathname}`;

  debug(
    `making ${(options.method || 'get').toUpperCase()} request to ${fullUrl} with headers: ${sanitizeHeaders(headers)}`,
  );

  return nodeFetch(fullUrl, {
    ...options,
    headers,
  }).then(res => {
    const warningHeader = res.headers.get('Warning');
    if (warningHeader) {
      debug(`received warning header: ${warningHeader}`);
      const warnings = parseWarningHeader(warningHeader);
      warnings.forEach(warning => {
        warn(warning.message, 'ReadMe API Warning:');
      });
    }
    return res;
  });
}

/**
 * Small handler for handling responses from our API.
 *
 * If we receive JSON errors, we throw an APIError exception.
 *
 * If we receive non-JSON responses, we consider them errors and throw them.
 *
 * @param rejectOnJsonError if omitted (or set to true), the function will return
 * an `APIError` if the JSON body contains an `error` property. If set to false,
 * the function will return a resolved promise containing the JSON object.
 *
 */
async function handleRes(res: Response, rejectOnJsonError = true) {
  const contentType = res.headers.get('content-type');
  const extension = mime.extension(contentType);
  if (extension === 'json') {
    const body = await res.json();
    debug(`received status code ${res.status} from ${res.url} with JSON response: ${JSON.stringify(body)}`);
    if (body.error && rejectOnJsonError) {
      return Promise.reject(new APIError(body));
    }
    return body;
  }
  if (res.status === SUCCESS_NO_CONTENT) {
    debug(`received status code ${res.status} from ${res.url} with no content`);
    return {};
  }
  // If we receive a non-JSON response, it's likely an error.
  // Let's debug the raw response body and throw it.
  const body = await res.text();
  debug(`received status code ${res.status} from ${res.url} with non-JSON response: ${body}`);
  return Promise.reject(body);
}

/**
 * Returns the basic auth header and any other defined headers for use in `node-fetch` API calls.
 *
 */
function cleanHeaders(key: string, inputHeaders: Headers = new Headers()) {
  const encodedKey = Buffer.from(`${key}:`).toString('base64');
  const headers = new Headers({
    Authorization: `Basic ${encodedKey}`,
  });

  for (const header of inputHeaders.entries()) {
    // If you supply `undefined` or `null` to the `Headers` API it'll convert that those to a
    // string.
    if (header[1] !== 'null' && header[1] !== 'undefined' && header[1].length > 0) {
      headers.set(header[0], header[1]);
    }
  }

  return headers;
}

export { cleanHeaders, getUserAgent, handleRes };
