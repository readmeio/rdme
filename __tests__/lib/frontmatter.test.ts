import type nock from 'nock';
import type { SchemaObject } from 'oas/types';

import { describe, it, expect, beforeEach, afterEach } from 'vitest';

import DocsUploadCommand from '../../src/commands/docs/upload.js';
import { fix } from '../../src/lib/frontmatter.js';
import { emptyMappings, fetchSchema } from '../../src/lib/readmeAPIFetch.js';
import { oasFetchMock } from '../helpers/get-api-mock.js';
import { setupOclifConfig } from '../helpers/oclif.js';

describe('#fix', () => {
  let command: DocsUploadCommand;
  let mock: nock.Scope;
  let schema: SchemaObject;

  beforeEach(async () => {
    const oclifConfig = await setupOclifConfig();
    command = new DocsUploadCommand([], oclifConfig);
    mock = oasFetchMock();
    schema = await fetchSchema.call(command);
  });

  afterEach(() => {
    mock.done();
  });

  it('should do nothing for an empty object', () => {
    const data = {};

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(false);
    expect(result.updatedData).toStrictEqual(data);
  });

  it('should do nothing for valid frontmatter', () => {
    const data = {
      title: 'Hello, world!',
      category: { uri: '/versions/stable/categories/guides/sup' },
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(false);
    expect(result.updatedData).toStrictEqual(data);
  });

  it('should do nothing for valid frontmatter (with invalid category uri)', () => {
    const data = {
      title: 'Hello, world!',
      category: { uri: 'sup' },
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(false);
    expect(result.updatedData).toStrictEqual(data);
  });

  it('should do nothing for valid frontmatter (with invalid parent uri)', () => {
    const data = {
      title: 'Hello, world!',
      parent: { uri: 'sup' },
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(false);
    expect(result.updatedData).toStrictEqual(data);
  });

  it('should fix legacy category id and use mappings', () => {
    const data = {
      title: 'Hello, world!',
      category: '5f92cbf10cf217478ba93561',
    };

    const result = fix.call(command, data, schema, {
      categories: { '5f92cbf10cf217478ba93561': 'some-slug' },
      parentPages: {},
    });

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "category": {
          "uri": "some-slug",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix legacy category id and use fallback mapping', () => {
    const data = {
      title: 'Hello, world!',
      category: '5f92cbf10cf217478ba93561',
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "category": {
          "uri": "uri-that-does-not-map-to-5f92cbf10cf217478ba93561",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix legacy category slug', () => {
    const data = {
      title: 'Hello, world!',
      categorySlug: 'some-slug',
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "category": {
          "uri": "some-slug",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix legacy parent page id and use mappings', () => {
    const data = {
      title: 'Hello, world!',
      parentDoc: '5f92cbf10cf217478ba93561',
    };

    const result = fix.call(command, data, schema, {
      categories: {},
      parentPages: { '5f92cbf10cf217478ba93561': 'some-slug' },
    });

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "parent": {
          "uri": "some-slug",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should delete legacy parent page id if no mapping is available', () => {
    const data = {
      title: 'Hello, world!',
      parentDoc: '5f92cbf10cf217478ba93561',
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix legacy parent page slug', () => {
    const data = {
      title: 'Hello, world!',
      parentDocSlug: 'some-slug',
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "parent": {
          "uri": "some-slug",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix excerpt', () => {
    const data = {
      title: 'Hello, world!',
      excerpt: 'This is an excerpt',
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "content": {
          "excerpt": "This is an excerpt",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix position', () => {
    const data = {
      title: 'Hello, world!',
      order: 5,
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "position": 5,
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix privacy (public)', () => {
    const data = {
      title: 'Hello, world!',
      hidden: false,
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "privacy": {
          "view": "public",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it('should fix privacy (anyone_with_link)', () => {
    const data = {
      title: 'Hello, world!',
      hidden: true,
    };

    const result = fix.call(command, data, schema, emptyMappings);

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "privacy": {
          "view": "anyone_with_link",
        },
        "title": "Hello, world!",
      }
    `);
  });

  it.todo('should fix metadata object');

  it.todo('should fix content.link object');

  it.todo('should fix content.next object');

  it('should fix multiple issues', () => {
    const data = {
      title: 'Hello, world!',
      category: '5f92cbf10cf217478ba93561',
      parentDocSlug: 'some-parent-slug',
      excerpt: 'This is an excerpt',
      content: {
        body: 'This is the body',
      },
      order: 7,
      hidden: true,
      slug: 'some-slug',
    };

    const result = fix.call(command, data, schema, {
      categories: { '5f92cbf10cf217478ba93561': 'some-category-slug' },
      parentPages: { '5f92cbf10cf217478ba93561': 'some-parent-slug' },
    });

    expect(result.hasIssues).toBe(true);
    expect(result.updatedData).toMatchInlineSnapshot(`
      {
        "category": {
          "uri": "some-category-slug",
        },
        "content": {
          "body": "This is the body",
          "excerpt": "This is an excerpt",
        },
        "parent": {
          "uri": "some-parent-slug",
        },
        "position": 7,
        "privacy": {
          "view": "anyone_with_link",
        },
        "slug": "some-slug",
        "title": "Hello, world!",
      }
    `);
  });
});
