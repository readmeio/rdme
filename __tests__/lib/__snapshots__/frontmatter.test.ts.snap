// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#fix (guides migrate) > should delete legacy parent page id if no mapping is available 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix excerpt 1`] = `
{
  "content": {
    "excerpt": "This is an excerpt",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix fullscreen 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix htmlmode (false) 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix htmlmode (true) 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix legacy category id and use fallback mapping 1`] = `
{
  "category": {
    "uri": "uri-that-does-not-map-to-5f92cbf10cf217478ba93561",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix legacy category id and use mappings 1`] = `
{
  "category": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix legacy category slug 1`] = `
{
  "category": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix legacy parent page id and use mappings 1`] = `
{
  "parent": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix legacy parent page slug 1`] = `
{
  "parent": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix multiple issues 1`] = `
{
  "category": {
    "uri": "some-category-slug",
  },
  "content": {
    "body": "This is the body",
    "excerpt": "This is an excerpt",
  },
  "parent": {
    "uri": "some-parent-slug",
  },
  "position": 7,
  "privacy": {
    "view": "anyone_with_link",
  },
  "slug": "some-slug",
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix position 1`] = `
{
  "position": 5,
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix privacy (anyone_with_link) 1`] = `
{
  "privacy": {
    "view": "anyone_with_link",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix privacy (anyone_with_link, \`hidden\` is a string) 1`] = `
{
  "privacy": {
    "view": "anyone_with_link",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix privacy (public) 1`] = `
{
  "privacy": {
    "view": "public",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides migrate) > should fix privacy (public, \`hidden\` is a string) 1`] = `
{
  "privacy": {
    "view": "public",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should delete legacy parent page id if no mapping is available 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix excerpt 1`] = `
{
  "content": {
    "excerpt": "This is an excerpt",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix fullscreen 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix htmlmode (false) 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix htmlmode (true) 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix legacy category id and use fallback mapping 1`] = `
{
  "category": {
    "uri": "uri-that-does-not-map-to-5f92cbf10cf217478ba93561",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix legacy category id and use mappings 1`] = `
{
  "category": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix legacy category slug 1`] = `
{
  "category": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix legacy parent page id and use mappings 1`] = `
{
  "parent": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix legacy parent page slug 1`] = `
{
  "parent": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix multiple issues 1`] = `
{
  "category": {
    "uri": "some-category-slug",
  },
  "content": {
    "body": "This is the body",
    "excerpt": "This is an excerpt",
  },
  "parent": {
    "uri": "some-parent-slug",
  },
  "position": 7,
  "privacy": {
    "view": "anyone_with_link",
  },
  "slug": "some-slug",
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix position 1`] = `
{
  "position": 5,
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix privacy (anyone_with_link) 1`] = `
{
  "privacy": {
    "view": "anyone_with_link",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix privacy (anyone_with_link, \`hidden\` is a string) 1`] = `
{
  "privacy": {
    "view": "anyone_with_link",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix privacy (public) 1`] = `
{
  "privacy": {
    "view": "public",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (guides upload) > should fix privacy (public, \`hidden\` is a string) 1`] = `
{
  "privacy": {
    "view": "public",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should delete legacy parent page id if no mapping is available 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix excerpt 1`] = `
{
  "content": {
    "excerpt": "This is an excerpt",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix fullscreen 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix htmlmode (false) 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix htmlmode (true) 1`] = `
{
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix legacy category id and use fallback mapping 1`] = `
{
  "category": {
    "uri": "uri-that-does-not-map-to-5f92cbf10cf217478ba93561",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix legacy category id and use mappings 1`] = `
{
  "category": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix legacy category slug 1`] = `
{
  "category": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix legacy parent page id and use mappings 1`] = `
{
  "parent": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix legacy parent page slug 1`] = `
{
  "parent": {
    "uri": "some-slug",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix multiple issues 1`] = `
{
  "category": {
    "uri": "some-category-slug",
  },
  "content": {
    "body": "This is the body",
    "excerpt": "This is an excerpt",
  },
  "parent": {
    "uri": "some-parent-slug",
  },
  "position": 7,
  "privacy": {
    "view": "anyone_with_link",
  },
  "slug": "some-slug",
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix position 1`] = `
{
  "position": 5,
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix privacy (anyone_with_link) 1`] = `
{
  "privacy": {
    "view": "anyone_with_link",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix privacy (anyone_with_link, \`hidden\` is a string) 1`] = `
{
  "privacy": {
    "view": "anyone_with_link",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix privacy (public) 1`] = `
{
  "privacy": {
    "view": "public",
  },
  "title": "Hello, world!",
}
`;

exports[`#fix (reference upload) > should fix privacy (public, \`hidden\` is a string) 1`] = `
{
  "privacy": {
    "view": "public",
  },
  "title": "Hello, world!",
}
`;
